### ISSUE TITLE
Cargo can't handle remote forced tag updates

### ISSUE THREAD

**Issue by mthebridge** [rust-lang/cargo#5109]
If using a git repo for dependencies, and a tag gets force-updated on the remote repo, cargo can't seem to ever be told to pull down the new version of the tag.

Essentially I had this in my `Cargo.toml`:

```toml
[dependencies]
foo = { git = "ssh://mygit.example.com/foo.git", tag = "1.2.3" }
```

Then tag 1.2.3 got force updated and I couldn't get it to update.

Running `cargo update -p ssh://mygit.example.com/foo.git#foo:1.2.3` just gives:

```
error: There are multiple `foo` packages in your project, and the specification `ssh://mygit.example.com/foo.git#foo:1.2.3` is ambiguous.
Please re-run this command with `-p <spec>` where `<spec>` is one of the following:
  ssh://mygit.example.com/foo.git#foo:1.2.3
  ssh://mygit.example.com/foo.git#foo:1.2.3
```
I couldn't find any way of getting cargo to distinguish between them. In the end I just blew away `~/.cargo/git/` and started again.  

I'm aware that force pushing tag updates to git remotes is bad practice.  But it occasionally happens and it would be nice to just have cargo resolve this on `cargo update`.

Using Cargo version: `cargo 0.25.0 (8c93e0895 2018-02-01)`

**Comment by alexcrichton** [comment]
Thanks for the report @mthebridge! This actually I think isn't necessarily related to tags but moreso related to how `cargo update` works. The `-p` flag is supposed to indicate "please update this node in the dependency graph" but it actually doesn't have syntax for differentiating these two dependencies:

```toml
[dependencies]
foo = { git = "ssh://mygit.example.com/foo.git }
bar = { git = "ssh://mygit.example.com/foo.git", tag = "1.2.3" }
```

So sort of what's happening here is that an existing bug is being exposed in Cargo where you aren't able to say "please update just this one node". @mthebridge I wonder though if you intended to use `[patch]` instead though to avoid compiling this repository twice?

**Comment by mthebridge** [comment]
Thanks for the speedy response @alexcrichton.  Actually it's weirder than I thought, blowing away `~/.cargo/git/` didn't help me.

I don't think `[patch]` is relevant here?  I don't want to compile the repo twice at all, I want to get rid of the old record of the tag entirely.  But that doesn't seem possible.

Here's what I did in more detail.
* I have my crate (let's call it `bar`) which depends on `foo`, which is pinned to a git tag.
* That `foo` tag had only just been released (by me in this case, I work on the dependent crate too) and I find it has a crass bug.  Nobody else is using the crate so I fix `foo` and force update the tag for speed.
* I want to move `bar` to the "fixed" version of `foo` - I still want it to depend on the same git tag, just the newer version.
* Nothing I do - even blowing away all cargo's cache and starting again - will persuade Cargo that the tag has moved.  In the end, I had to manually replace the old hash with the new in `foo`'s `Cargo.lock`.

The only other solution I could think of would be to set `foo`'s` Cargo.toml` to use the specific git hash, rather than the tag, but that's a bit ugly.

It's entirely possible this is all as designed and I've missed something!
